### ISSUE TITLE
Issue with latest stable version. KeyError: 'CommandRcon'

### ISSUE THREAD

**Issue by FliesWithWind** [Yoshi-E/ArmaRconDiscordBot#58]
Hi,

I did an update to latest stable version and I'm getting an error after log pre-scanning.
```
Traceback (most recent call last):
  File "/home/game1/ArmaRconDiscordBot/modules/arma/module.py", line 81, in on_ready
    self.CommandRcon = self.bot.cogs["CommandRcon"]
KeyError: 'CommandRcon'
```
Is the branch all up to date?

Thanks for the help in advance. 
Attaching more output that I've got.

```
2022-09-24 14:54:00 WARNING  discord.gateway Shard ID None heartbeat blocked for more than 340 seconds.
Loop thread traceback (most recent call last):
  File "/home/game1/ArmaRconDiscordBot/bot.py", line 69, in <module>
    main()
  File "/home/game1/ArmaRconDiscordBot/bot.py", line 58, in main
    bot.run(cfg["TOKEN"])
  File "/home/game1/.local/lib/python3.10/site-packages/discord/client.py", line 828, in run
    asyncio.run(runner())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 633, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.10/asyncio/base_events.py", line 600, in run_forever
    self._run_once()
  File "/usr/lib/python3.10/asyncio/base_events.py", line 1896, in _run_once
    handle._run()
  File "/usr/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/game1/.local/lib/python3.10/site-packages/discord/ext/commands/bot.py", line 1012, in load_extension
    await self._load_from_module_spec(spec, name)
  File "/home/game1/.local/lib/python3.10/site-packages/discord/ext/commands/bot.py", line 946, in _load_from_module_spec
    await setup(self)
  File "/home/game1/ArmaRconDiscordBot/modules/arma/module.py", line 346, in setup
    await bot.add_cog(CommandArma(bot))
  File "/home/game1/ArmaRconDiscordBot/modules/arma/module.py", line 44, in __init__
    self.readLog.pre_scan()
  File "/home/game1/ArmaRconDiscordBot/modules/arma/readLog.py", line 212, in pre_scan
    self.scanfile(_log)
  File "/home/game1/ArmaRconDiscordBot/modules/arma/readLog.py", line 314, in scanfile
    self.processLogLine(line)
  File "/home/game1/ArmaRconDiscordBot/modules/arma/readLog.py", line 230, in processLogLine
    timestamp, msg = self.splitTimestamp(line)
  File "/home/game1/ArmaRconDiscordBot/modules/arma/readLog.py", line 296, in splitTimestamp
    m = re.match(r"^\s?([0-9]{1,2}:[0-9]{2}:[0-9]{2})\s(.*)", log_line)
  File "/usr/lib/python3.10/re.py", line 190, in match
    return _compile(pattern, flags).match(string)

Traceback (most recent call last):
  File "/home/game1/ArmaRconDiscordBot/modules/arma/module.py", line 81, in on_ready
    self.CommandRcon = self.bot.cogs["CommandRcon"]
KeyError: 'CommandRcon'

'CommandRcon'
2022-09-24 14:54:01 ERROR    discord.client Ignoring exception in on_ready
Traceback (most recent call last):
  File "/home/game1/.local/lib/python3.10/site-packages/discord/client.py", line 409, in _run_event
    await coro(*args, **kwargs)
  File "/home/game1/ArmaRconDiscordBot/bot.py", line 45, in on_ready
    roles += await guild.fetch_roles()
  File "/home/game1/.local/lib/python3.10/site-packages/discord/guild.py", line 3069, in fetch_roles
    data = await self._state.http.get_roles(self.id)
  File "/home/game1/.local/lib/python3.10/site-packages/discord/http.py", line 624, in request
    async with self.__session.request(method, url, **kwargs) as response:
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/client.py", line 1141, in __aenter__
    self._resp = await self._coro
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/client.py", line 560, in _request
    await resp.start(conn)
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 894, in start
    with self._timer:
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/helpers.py", line 720, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
2022-09-24 14:54:02 INFO     discord.gateway Shard ID None session has been invalidated.
2022-09-24 14:54:07 INFO     discord.gateway Shard ID None has connected to Gateway (Session ID: 66dac5bd1656c7e94b35eaf2fd547599).
Logged in as Liberation [817030019842834482]
<discord.utils.SequenceProxy object at 0x7f8a695f8b50>
------------
Task exception was never retrieved
future: <Task finished name='Task-26' coro=<BotBase.load_extension() done, defined at /home/game1/.local/lib/python3.10/site-packages/discord/ext/commands/bot.py:961> exception=ExtensionAlreadyLoaded("Extension 'modules.arma.module' is already loaded.")>
Traceback (most recent call last):
  File "/home/game1/.local/lib/python3.10/site-packages/discord/ext/commands/bot.py", line 1006, in load_extension
    raise errors.ExtensionAlreadyLoaded(name)
discord.ext.commands.errors.ExtensionAlreadyLoaded: Extension 'modules.arma.module' is already loaded.
Task exception was never retrieved
future: <Task finished name='Task-27' coro=<BotBase.load_extension() done, defined at /home/game1/.local/lib/python3.10/site-packages/discord/ext/commands/bot.py:961> exception=ExtensionAlreadyLoaded("Extension 'modules.core.module' is already loaded.")>
Traceback (most recent call last):
  File "/home/game1/.local/lib/python3.10/site-packages/discord/ext/commands/bot.py", line 1006, in load_extension
    raise errors.ExtensionAlreadyLoaded(name)
discord.ext.commands.errors.ExtensionAlreadyLoaded: Extension 'modules.core.module' is already loaded.
Task exception was never retrieved
future: <Task finished name='Task-28' coro=<BotBase.load_extension() done, defined at /home/game1/.local/lib/python3.10/site-packages/discord/ext/commands/bot.py:961> exception=ExtensionAlreadyLoaded("Extension 'modules.errorhandle.module' is already loaded.")>
Traceback (most recent call last):
  File "/home/game1/.local/lib/python3.10/site-packages/discord/ext/commands/bot.py", line 1006, in load_extension
    raise errors.ExtensionAlreadyLoaded(name)
discord.ext.commands.errors.ExtensionAlreadyLoaded: Extension 'modules.errorhandle.module' is already loaded.
Exiting...
Exiting... A
Bot Terminated (KeyboardInterrupt)
```

**Comment by Yoshi-E** [comment]
Some debug code made it into live it seems...
I apologize for the oversight.

https://github.com/Yoshi-E/ArmaRconDiscordBot/blob/ee47f9619c537649eafb4436f35e9961194600b4/bot.py#L47-L54

Already removed it from the dev branch.

However the problem might be a different one too.
The recent changes to the discord API changed how modules are loaded. Now they are no longer loaded in sequential order but loaded async in runtime.
This is problematic as they modules depend on each other to be loaded once the bot is started.
In this case its expecting the CommandRcon module to be loaded, but it wasnt in time.

Will work on a fix.

**Comment by Yoshi-E** [comment]
Added the new setup hook they added to hopefully fix the loading be done in before the bot is ready.
#59 94b668334cd0ca6448e480846c5a918282e9f490

**Comment by FliesWithWind** [comment]
It seems there are still some issues present.

```
Log file path: '/home/game1/ArmaRconDiscordBot/modules/core/error.log'
2022-09-25 15:24:38 INFO     discord.client logging in using static token
[WARNING] Port '8000' already in use, using '8001' instead.
Settings page online on: http://localhost:8001/
Pre-scanning: arma3server-console.log

2022-09-25 15:36:06 ERROR    discord.client Attempting a reconnect in 1.83s
Traceback (most recent call last):
  File "/home/game1/.local/lib/python3.10/site-packages/discord/gateway.py", line 371, in from_client
    socket = await client.http.ws_connect(url)
  File "/home/game1/.local/lib/python3.10/site-packages/discord/http.py", line 536, in ws_connect
    return await self.__session.ws_connect(url, **kwargs)
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/client.py", line 779, in _ws_connect
    resp = await self.request(
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/client.py", line 536, in _request
    conn = await self._connector.connect(
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/connector.py", line 540, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/connector.py", line 901, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/connector.py", line 1152, in _create_direct_connection
    hosts = await asyncio.shield(host_resolved)
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/game1/.local/lib/python3.10/site-packages/discord/client.py", line 624, in connect
    self.ws = await asyncio.wait_for(coro, timeout=60.0)
  File "/usr/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
Traceback (most recent call last):
  File "/home/game1/.local/lib/python3.10/site-packages/discord/gateway.py", line 371, in from_client
    socket = await client.http.ws_connect(url)
  File "/home/game1/.local/lib/python3.10/site-packages/discord/http.py", line 536, in ws_connect
    return await self.__session.ws_connect(url, **kwargs)
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/client.py", line 779, in _ws_connect
    resp = await self.request(
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/client.py", line 536, in _request
    conn = await self._connector.connect(
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/connector.py", line 540, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/connector.py", line 901, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
  File "/home/game1/.local/lib/python3.10/site-packages/aiohttp/connector.py", line 1152, in _create_direct_connection
    hosts = await asyncio.shield(host_resolved)
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/game1/.local/lib/python3.10/site-packages/discord/client.py", line 624, in connect
    self.ws = await asyncio.wait_for(coro, timeout=60.0)
  File "/usr/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/game1/ArmaRconDiscordBot/bot.py", line 67, in <module>
    main()
  File "/home/game1/ArmaRconDiscordBot/bot.py", line 56, in main
    bot.run(cfg["TOKEN"])
  File "/home/game1/.local/lib/python3.10/site-packages/discord/client.py", line 828, in run
    asyncio.run(runner())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/home/game1/.local/lib/python3.10/site-packages/discord/client.py", line 817, in runner
    await self.start(token, reconnect=reconnect)
  File "/home/game1/.local/lib/python3.10/site-packages/discord/client.py", line 746, in start
    await self.connect(reconnect=reconnect)
  File "/home/game1/.local/lib/python3.10/site-packages/discord/client.py", line 684, in connect
    sequence=self.ws.sequence,
AttributeError: 'NoneType' object has no attribute 'sequence'
```

**Comment by Yoshi-E** [comment]
This seems unrelated to my bot though.

Could you provide the output of pip freeze?

Its probably cause by some version of the discord.py module

**Comment by FliesWithWind** [comment]
Here's the output of pip freeze. Updating aiohttp did not help :(

```
aiohttp==3.8.3
aiosignal==1.2.0
async-timeout==4.0.2
attrs==19.3.0
Automat==0.8.0
bec-rcon @ git+https://github.com/Yoshi-E/Python-BEC-RCon@dd18f441cafe54586b4b930ed9efba992cf165c0
blinker==1.4
certifi==2019.11.28
chardet==3.0.4
charset-normalizer==2.1.1
Click==7.0
colorama==0.4.3
command-not-found==0.3
configobj==5.0.6
constantly==15.1.0
contourpy==1.0.5
cryptography==2.8
cupshelpers==1.0
cycler==0.11.0
dbus-python==1.2.16
defer==1.0.6
discord==2.0.0
discord.py==2.0.1
distro==1.4.0
distro-info===0.23ubuntu1
entrypoints==0.3
fonttools==4.37.3
frozenlist==1.3.1
func-timeout==4.3.5
geoip2==4.6.0
httplib2==0.14.0
hyperlink==19.0.0
idna==2.8
incremental==16.10.1
iniconfig==1.1.1
keyring==18.0.1
kiwisolver==1.4.4
language-selector==0.1
launchpadlib==1.10.13
lazr.restfulclient==0.14.2
lazr.uri==1.0.3
macaroonbakery==1.3.1
matplotlib==3.6.0
maxminddb==2.2.0
multidict==6.0.2
netifaces==0.10.4
numpy==1.23.3
oauthlib==3.1.0
packaging==21.3
pexpect==4.6.0
Pillow==9.2.0
pluggy==1.0.0
prettytable==3.4.1
protobuf==3.6.1
psutil==5.9.2
py==1.11.0
pyasn1==0.4.2
pyasn1-modules==0.2.1
pycairo==1.16.2
pycups==1.9.73
pycurl==7.43.0.2
pygeoip==0.3.2
PyGObject==3.36.0
PyHamcrest==1.9.0
PyJWT==1.7.1
pymacaroons==0.13.0
PyNaCl==1.3.0
pyOpenSSL==19.0.0
pyparsing==3.0.9
pyRFC3339==1.1
pytest==7.1.3
pytest-timeout==2.1.0
python-apt==2.0.0+ubuntu0.20.4.8
python-dateutil==2.8.2
python-debian===0.1.36ubuntu1
pytz==2019.3
PyYAML==5.3.1
requests==2.28.1
requests-unixsocket==0.2.0
scipy==1.9.1
SecretStorage==2.3.1
service-identity==18.1.0
simplejson==3.16.0
six==1.14.0
sos==4.3
ssh-import-id==5.10
systemd-python==234
tomli==2.0.1
Twisted==18.9.0
ubuntu-advantage-tools==27.10
ufw==0.36
unattended-upgrades==0.1
urllib3==1.25.8
wadllib==1.3.3
wcwidth==0.2.5
yarl==1.8.1
zope.interface==4.7.1
```

**Comment by Yoshi-E** [comment]
I tested bot; the default install and your environment on both Ubuntu and Win10, but I could not recreate the error.

It is very likely related to an underlying Network issue as I suspected initially.
The error raise here is by the aiohttp module and the error is that the host can not be resolved, or some similar error. 
Sadly the error isnt quite specific.

If you are using docker it might be a bad dns configuration.
